<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scdneyDiseasePrediction">
<title>Disease outcome classification in Single Cell Patient Data Analysis • scdneyDiseasePrediction</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Disease outcome classification in Single Cell Patient Data Analysis">
<meta property="og:description" content="scdneyDiseasePrediction">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scdneyDiseasePrediction</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/disease_outcome_classification_schulte.html">Disease outcome classification in Single Cell Patient Data Analysis</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Disease outcome classification in Single Cell Patient Data Analysis</h1>
                        <h4 data-toc-skip class="author">Yue Cao<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:yue.cao@sydney.edu.au" class="email"&gt;yue.cao@sydney.edu.au&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a> Andy
Tran<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="mailto:andy.t@sydney.edu.au" class="email"&gt;andy.t@sydney.edu.au&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a>
</h4>
            
            <h4 data-toc-skip class="date">2022-10-17</h4>
      
      
      <div class="d-none name"><code>disease_outcome_classification_schulte.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>As single-cell technology advances, the number of multi-condition
multi-sample datasets increases. This workshop will discuss the
challenges and analytical focus associated with disease outcome
prediction using single cell data, which is one of the analytics
involved in such datasets. The workshop’s overall learning goal is to
investigate one possible approach to this task. We will also talk about
general analytic strategies and the critical thinking questions that
arise in the workflow.</p>
<br><br>
</div>
<div class="aimbox">
<div class="section level3">
<h3 id="preparation-and-assumed-knowledge">
<span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-pencil-alt fa-stack-1x fa-inverse"></i></span>
Preparation and assumed knowledge<a class="anchor" aria-label="anchor" href="#preparation-and-assumed-knowledge"></a>
</h3>
<ul>
<li>Knowledge of R syntax</li>
<li>Familiarity with the <a href="https://satijalab.org/seurat/" class="external-link">Seurat
class</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="learning-objectives">
<span class="fa-stack fa"><i class="fa fa-circle fa-stack-2x"></i><i class="fa fa-location-arrow fa-stack-1x fa-inverse"></i></span>
Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>Explore various strategies for disease outcome prediction using
single cell data<br>
</li>
<li>Understand the transformation from cell level features to patient
level features<br>
</li>
<li>Generate patient representations from gene expression matrix<br>
</li>
<li>Understand the characteristics of good classification models<br>
</li>
<li>Perform disease outcome prediction using the feature representation
and robust classification framework</li>
</ul>
</div>
</div>
<p><br><br></p>
<div class="section level3">
<h3 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<p>Structure of the 2-hour workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction + loading data</td>
<td>20m</td>
</tr>
<tr class="even">
<td>Extracting features from single cell data</td>
<td>40m</td>
</tr>
<tr class="odd">
<td>Break</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Perform disease outcome classification</td>
<td>40m</td>
</tr>
<tr class="odd">
<td>Concluding remarks</td>
<td>10m</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="introduction">1. Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The rise of single-cell or near single-cell resolution omics
technologies (e.g. spatial transcriptomics) has enabled the discovery of
cell- and cell type specific knowledge and have transformed our
understanding of biological systems. Because of the high-dimensionality
and complexity, over 1000 tools have been developed to extract
meaningful information from the high feature dimensions and uncover
biological insights. For example, our previous workshop focuses on
characterising the identity the state of cells and the relationship
between cells along a trajectory.</p>
<p>While these tools enable characterisation of individual cells, there
is a lack of tools that characterise individual samples as a whole based
on their cellular properties and investigate how these cellular
properties are driving disease outcomes. With the recent surge of
multi-condition and multi-sample single-cell studies, the question
becomes how do we represent cellular properties at the sample
(e.g. individual patient) level for linking such information with the
disease outcome and performing downstream analysis such as disease
outcome prediction.</p>
<p>In this workshop, we will demonstrate our approach for generating a
molecular representation for individual samples and using the
representation for a downstream application of disease outcome
classification.</p>
</div>
<div class="section level2">
<h2 id="loading-libraries-and-the-data">2. Loading libraries and the data<a class="anchor" aria-label="anchor" href="#loading-libraries-and-the-data"></a>
</h2>
<div class="section level3">
<h3 id="load-libraries">2.1 Load libraries<a class="anchor" aria-label="anchor" href="#load-libraries"></a>
</h3>
<p>First, load all the libraries we will be using in this workshop.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">thispackage</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ggplot2"</span>,  <span class="st">"ggthemes"</span> , <span class="st">"Seurat"</span> , <span class="st">"dplyr"</span> , <span class="st">"devtools"</span>,</span>
<span>                      <span class="st">"RCurl"</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="op">!</span> <span class="va">thispackage</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span> <span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="va">thispackage</span> <span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jrnold/ggthemes" class="external-link">ggthemes</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RCurl</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="st">"scFeatures"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span> <span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"SydneyBioX/scFeatures"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scFeatures</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="st">"ClassifyR"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span> <span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>     <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">}</span> </span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ClassifyR"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ClassifyR</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2022</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="loading-the-preprocessed-data">2.2 Loading the preprocessed data<a class="anchor" aria-label="anchor" href="#loading-the-preprocessed-data"></a>
</h3>
<p>We will use a single-cell RNA-sequencing (scRNA-seq) data of COVID-19
patients for this workshop. The dataset is taken from <a href="https://pubmed.ncbi.nlm.nih.gov/32810438/" class="external-link">Schulte-Schrepping et
al. 2020</a>. We have subsampled 20 mild and 20 severe patient samples
from this dataset. The original data can be accessed from the European
Genome-phenome Archive (EGA) with the accession number
EGAS00001004571.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"http://www.dropbox.com/s/vhnwch3362dlwr5/Schulte-schrepping_20_severe_20_mild.rds?raw=1"</span> , destfile <span class="op">=</span> <span class="st">"toy_data/Schulte-schrepping_20_severe_20_mild.rds"</span> , method<span class="op">=</span><span class="st">"auto"</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"toy_data/Schulte-schrepping_20_severe_20_mild.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualising-the-data">2.3 Visualising the data<a class="anchor" aria-label="anchor" href="#visualising-the-data"></a>
</h3>
<p>We can visualise the data using dimensionality reduction approaches
and colour the individual cells by the severity.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span> <span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">data</span> <span class="op">)</span>  </span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">data</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">data</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 37210</span></span>
<span><span class="co">#&gt; Number of edges: 1210153</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.9436</span></span>
<span><span class="co">#&gt; Number of communities: 26</span></span>
<span><span class="co">#&gt; Elapsed time: 7 seconds</span></span>
<span><span class="va">data</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">data</span> , dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">data</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"meta_severity"</span><span class="op">)</span></span></code></pre></div>
<p><img src="disease_outcome_classification_schulte_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i>
<i class="far fa-chart-bar fa-stack-1x fa-inverse"></i> </span>
<strong>Interpretation:</strong></p>
<p>Are the cells from mild/moderate and severe/critical patients easy or
difficult to distinguish?</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="how-can-i-characterise-each-individual-as-a-whole-from-the-matrix-of-genes-x-cells">3. How can I characterise each individual as a whole from the matrix
of genes x cells ?<a class="anchor" aria-label="anchor" href="#how-can-i-characterise-each-individual-as-a-whole-from-the-matrix-of-genes-x-cells"></a>
</h2>
<p>In this workshop, we use <a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btac590/6678979" class="external-link">scFeatures</a>
to generate molecular representation for each patient. The molecular
representation is interpretable and hence facilitates downstream
analysis of the patient. Overall, scFeatures generates features across
six categories representing different molecular views of cellular
characteristics. These include:<br>
- i) cell type proportions<br>
- ii) cell type specific gene expressions<br>
- iii) cell type specific pathway expressions<br>
- iv) cell type specific cell-cell interaction (CCI) scores<br>
- v) overall aggregated gene expressions<br>
- vi) spatial metrics<br>
The different types of features constructed enable a more comprehensive
multi-view understanding of each patient from a matrix of genes x
cells.</p>
<p><img src="../../../../../Dropbox%20(Sydney%20Uni)/Yue%20Cao%20Home%20Content/Github_push/scdneyDieasePrediction/articles/scFeatures_overview.png"></p>
<div class="section level3">
<h3 id="checking-the-data">3.1 Checking the data<a class="anchor" aria-label="anchor" href="#checking-the-data"></a>
</h3>
<p>Inspect the number of cells in each patient sample, as well as the
number of cells in each cell type.</p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<p>Is there any sample or cell types you should remove from the
data?</p>
</div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;       BN-03_FrozenPBMC_03.04.2020-FrozenPBMC </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt; BN-04_FreshEryLysis_23.04.2020-FreshEryLysis </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt; BN-05_FreshEryLysis_23.04.2020-FreshEryLysis </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt; BN-06_FreshEryLysis_23.04.2020-FreshEryLysis </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt; BN-08_FreshEryLysis_29.05.2020-FreshEryLysis </span></span>
<span><span class="co">#&gt;                                          994 </span></span>
<span><span class="co">#&gt; BN-11_FreshEryLysis_23.04.2020-FreshEryLysis </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt; BN-12_FreshEryLysis_14.05.2020-FreshEryLysis </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt; BN-12_FreshEryLysis_18.05.2020-FreshEryLysis </span></span>
<span><span class="co">#&gt;                                          410 </span></span>
<span><span class="co">#&gt; BN-18_FreshEryLysis_12.05.2020-FreshEryLysis </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt; BN-18_FreshEryLysis_14.05.2020-FreshEryLysis </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt; BN-18_FreshEryLysis_18.05.2020-FreshEryLysis </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt; BN-19_FreshEryLysis_12.05.2020-FreshEryLysis </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt; BN-19_FreshEryLysis_14.05.2020-FreshEryLysis </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                        C19-CB-0001_d11-Fresh </span></span>
<span><span class="co">#&gt;                                          970 </span></span>
<span><span class="co">#&gt;                       C19-CB-0001_d16-Frozen </span></span>
<span><span class="co">#&gt;                                          273 </span></span>
<span><span class="co">#&gt;                        C19-CB-0001_d7-Frozen </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                        C19-CB-0002_d13-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                        C19-CB-0002_d8-Frozen </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                        C19-CB-0003_d13-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                       C19-CB-0003_d18-Frozen </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                       C19-CB-0005_d15-Frozen </span></span>
<span><span class="co">#&gt;                                          448 </span></span>
<span><span class="co">#&gt;                       C19-CB-0005_d20-Frozen </span></span>
<span><span class="co">#&gt;                                          875 </span></span>
<span><span class="co">#&gt;                        C19-CB-0008_d13-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                        C19-CB-0008_d20-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                        C19-CB-0009_d16-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                         C19-CB-0009_d9-Fresh </span></span>
<span><span class="co">#&gt;                                          938 </span></span>
<span><span class="co">#&gt;                        C19-CB-0011_d13-Fresh </span></span>
<span><span class="co">#&gt;                                          932 </span></span>
<span><span class="co">#&gt;                        C19-CB-0012_d16-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                         C19-CB-0012_d9-Fresh </span></span>
<span><span class="co">#&gt;                                          741 </span></span>
<span><span class="co">#&gt;                        C19-CB-0013_d15-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                         C19-CB-0013_d8-Fresh </span></span>
<span><span class="co">#&gt;                                          629 </span></span>
<span><span class="co">#&gt;                        C19-CB-0016_d16-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                        C19-CB-0020_d16-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                        C19-CB-0021_d18-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                       C19-CB-0052_d10-Frozen </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                        C19-CB-0053_d3-Frozen </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                        C19-CB-0198_d18-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                         C19-CB-0199_d8-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                        C19-CB-0204_d15-Fresh </span></span>
<span><span class="co">#&gt;                                         1000 </span></span>
<span><span class="co">#&gt;                         C19-CB-0214_d7-Fresh </span></span>
<span><span class="co">#&gt;                                         1000</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">celltype</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;            B    CD14 Mono    CD16 Mono        CD4 T        CD8 T           DC </span></span>
<span><span class="co">#&gt;         1863         7241          919         7612         1763          260 </span></span>
<span><span class="co">#&gt;           DN          gdT         HSPC          ILC intermediate         MAIT </span></span>
<span><span class="co">#&gt;           52          264           43            1          645          328 </span></span>
<span><span class="co">#&gt;         MAST   Neutrophil           NK          NKT       Plasma     Platelet </span></span>
<span><span class="co">#&gt;           60         9853         3518         1561          513          404 </span></span>
<span><span class="co">#&gt;          RBC   unassigned </span></span>
<span><span class="co">#&gt;            2          308</span></span></code></pre></div>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<p>After running the <code>process_data</code> function, are there still
any patient samples or cell types that you should remove from the
data?</p>
</div>
</div>
<div class="section level3">
<h3 id="creating-molecular-representations-of-patients">3.1 Creating molecular representations of patients<a class="anchor" aria-label="anchor" href="#creating-molecular-representations-of-patients"></a>
</h3>
<p>All the feature types can be generated in one line of code. This runs
the function using default settings for all parameters, for more
information, type <code><a href="https://rdrr.io/pkg/scFeatures/man/scFeatures.html" class="external-link">?scFeatures</a></code>.</p>
<p>Given that this step may take up to 20 minutes, we have already
generated the result and saved it in the
<code>intermediate_result</code> folder. You could skip this step and
proceed to the next step for the purposes of this workshop.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Here we label the samples using the severity, such as it is easier later on to retrieve the severity outcomes of the samples. </span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">sample</span>, <span class="st">"_cond_"</span> , <span class="va">data</span><span class="op">$</span><span class="va">meta_severity</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scFeatures/man/scFeatures.html" class="external-link">scFeatures</a></span><span class="op">(</span><span class="va">data</span>, ncores <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># setwd("intermediate_result/")</span></span>
<span><span class="co"># saveRDS( scfeatures_result , "scfeatures_result_wilk_full.rds")</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualising-and-exploring-scfeatures-output">3.2 Visualising and exploring scFeatures output<a class="anchor" aria-label="anchor" href="#visualising-and-exploring-scfeatures-output"></a>
</h3>
<p>We have generated a total of 13 feature types and stored them in a
list. All generated feature types are stored in a matrix of samples by
features.<br>
For example, the first list element contains the feature type
“proportion_raw”, which contains the cell type proportion features for
each patient sample. We could print out the first 5 columns and first 5
rows of the first element to see.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://www.dropbox.com/s/volcbtmcz4nlqw4/scfeatures_result_schulte.rds?raw=1"</span> , destfile <span class="op">=</span> <span class="st">"intermediate_result/scfeatures_result_schulte.rds"</span> , method<span class="op">=</span><span class="st">"auto"</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">scfeatures_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"intermediate_result/scfeatures_result_schulte.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we have generated a total of 13 feature types</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "proportion_raw"       "proportion_logit"     "proportion_ratio"    </span></span>
<span><span class="co">#&gt;  [4] "gene_mean_celltype"   "gene_prop_celltype"   "gene_cor_celltype"   </span></span>
<span><span class="co">#&gt;  [7] "pathway_gsva"         "pathway_mean"         "pathway_prop"        </span></span>
<span><span class="co">#&gt; [10] "CCI"                  "gene_mean_aggregated" "gene_prop_aggregated"</span></span>
<span><span class="co">#&gt; [13] "gene_cor_aggregated"</span></span>
<span></span>
<span><span class="co"># each row is a sample, each column is a feature </span></span>
<span><span class="va">scfeatures_result</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                                                   B CD14 Mono  CD16 Mono</span></span>
<span><span class="co">#&gt; C19-CB-0001_d11-Fresh_cond_Mild/Moderate  0.2033024 0.4210526 0.02373581</span></span>
<span><span class="co">#&gt; C19-CB-0003_d13-Fresh_cond_Mild/Moderate  0.1524574 0.5687061 0.06218656</span></span>
<span><span class="co">#&gt; C19-CB-0002_d13-Fresh_cond_Mild/Moderate  0.2312312 0.2732733 0.03203203</span></span>
<span><span class="co">#&gt; C19-CB-0002_d8-Frozen_cond_Mild/Moderate  0.0670000 0.2780000 0.03000000</span></span>
<span><span class="co">#&gt; C19-CB-0001_d16-Frozen_cond_Mild/Moderate 0.2271062 0.3333333 0.04395604</span></span>
<span><span class="co">#&gt;                                                CD4 T      CD8 T</span></span>
<span><span class="co">#&gt; C19-CB-0001_d11-Fresh_cond_Mild/Moderate  0.12899897 0.02270382</span></span>
<span><span class="co">#&gt; C19-CB-0003_d13-Fresh_cond_Mild/Moderate  0.02407222 0.02407222</span></span>
<span><span class="co">#&gt; C19-CB-0002_d13-Fresh_cond_Mild/Moderate  0.06006006 0.05205205</span></span>
<span><span class="co">#&gt; C19-CB-0002_d8-Frozen_cond_Mild/Moderate  0.04600000 0.06200000</span></span>
<span><span class="co">#&gt; C19-CB-0001_d16-Frozen_cond_Mild/Moderate 0.11721612 0.02564103</span></span></code></pre></div>
<p>Once the features are generated, you may wish to visually explore the
features. For example, with cell-cell communication, a heatmap
summarising the number of interactions between cell types would provide
more direct insight than a matrix of interaction scores.</p>
<p>To accommodate for this need, scFeatures contains a function
<code>run_association_study_report</code> that enables the user to
readily visualise and explore all generated features with one line of
code.</p>
<p>Note, because this function knits a rmd file, please open a new R
script, paste the command in and run it there.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify a folder to store the html report. Here we store it in the current working directory. </span></span>
<span><span class="va">output_folder</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scFeatures/man/run_association_study_report.html" class="external-link">run_association_study_report</a></span><span class="op">(</span><span class="va">scfeatures_result</span>, <span class="va">output_folder</span> <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="are-the-generated-features-sensible">3.3 Are the generated features sensible?<a class="anchor" aria-label="anchor" href="#are-the-generated-features-sensible"></a>
</h3>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<p>Using the HTML, we can look at some of the critical thinking
questions that a researcher would ask about the generated features.
These questions are exploratory and there is no right or wrong
answer.</p>
<ul>
<li>Do the generated features look reasonable?<br>
</li>
<li>Which cell type(s) would you like to focus on at your next stage of
analysis?<br>
</li>
<li>Which feature type(s) would you like to focus on at your next stage
of analysis?<br>
</li>
<li>Are the conditions in your data relatively easy or difficult to
distinguish?</li>
</ul>
</div>
</div>
</div>
<div class="section level2">
<h2 id="how-can-i-perform-disease-outcome-classification-using-the-molecular-representation-of-patients">4. How can I perform disease outcome classification using the
molecular representation of patients?<a class="anchor" aria-label="anchor" href="#how-can-i-perform-disease-outcome-classification-using-the-molecular-representation-of-patients"></a>
</h2>
<p>Now that we have generated the patient representation, in this
section we will examine a useful case study of using the representation
to perform disease outcome classification.</p>
<div class="section level3">
<h3 id="building-classification-model">4.1 Building classification model<a class="anchor" aria-label="anchor" href="#building-classification-model"></a>
</h3>
<p>In this workshop we use the <a href="https://academic.oup.com/bioinformatics/article/31/11/1851/2365648?login=false" class="external-link">classifyR</a>
package to build a classification model. It provides an implementation
of a typical framework for classification, including a function that
performs repeated cross-validation with one line of code.</p>
<p><img src="../../../../../Dropbox%20(Sydney%20Uni)/Yue%20Cao%20Home%20Content/Github_push/scdneyDieasePrediction/articles/classifyR_overview.png"></p>
<p>We will build a classification model for each feature type.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># First clean the column names</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\s|[[:punct:]]"</span>, <span class="st">"."</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># obtain the patient outcome, which is stored in the rownames of each matrix </span></span>
<span><span class="va">outcome</span> <span class="op">=</span> <span class="va">scfeatures_result</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">rownames</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="st">"_cond_"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Recall in the previous section that we have stored the 13 feature
types matrix in a list. Instead of retrieving each matrix from the list,
classifyR can directly take a list of matrices as an input and run
repeated cross-validation model on each matrix individually.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">classifyr_result</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ClassifyR/man/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span><span class="va">scfeatures_result</span>,</span>
<span>                            <span class="va">outcome</span>, </span>
<span>                            classifier <span class="op">=</span> <span class="st">"DLDA"</span>,</span>
<span>                            nCores <span class="op">=</span> <span class="fl">10</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># because this can take a while, save for future use</span></span>
<span><span class="co"># saveRDS("intermediate_result/classifyr_result.rds)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualising-the-classification-performance">4.2 Visualising the classification performance<a class="anchor" aria-label="anchor" href="#visualising-the-classification-performance"></a>
</h3>
<p>To examine the classification model performance, we first need to
specify a metric to calculate. Here, we calculate the balanced
accuracy.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://www.dropbox.com/s/4018lih0qxrgg96/classifyr_result_DLDA.rds?raw=1"</span> , destfile <span class="op">=</span><span class="st">"intermediate_result/classifyr_result_DLDA.rds"</span> , method<span class="op">=</span><span class="st">"auto"</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">classifyr_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"intermediate_result/classifyr_result_DLDA.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">classifyr_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">classifyr_result</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/ClassifyR/man/calcPerformance.html" class="external-link">calcCVperformance</a></span><span class="op">(</span><span class="va">x</span>, performanceType <span class="op">=</span> <span class="st">"Balanced Accuracy"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Format the output and visualise the accuracy using boxplots.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">accuracy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">classifyr_result</span>, <span class="va">performance</span><span class="op">)</span> <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">accuracy</span> <span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span> <span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">classifyr_result</span><span class="op">)</span> , <span class="st">"\\."</span> <span class="op">)</span>, <span class="va">`[`</span> , <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">accuracy</span> <span class="op">&lt;-</span> <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">accuracy</span><span class="op">)</span></span>
<span></span>
<span><span class="va">accuracy_DLDA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">accuracy</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">variable</span>, y <span class="op">=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Feature types"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Balanced Accuracy - DLDA"</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">accuracy_DLDA</span></span></code></pre></div>
<p><img src="disease_outcome_classification_schulte_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i>
<i class="far fa-chart-bar fa-stack-1x fa-inverse"></i> </span>
<strong>Interpretation:</strong></p>
<p>Based on the classification performance, which feature type would you
like to focus on at your next stage of analysis?</p>
</div>
</div>
<div class="section level3">
<h3 id="what-features-are-important-for-disease-outcome-prediction">4.3 What features are important for disease outcome prediction?<a class="anchor" aria-label="anchor" href="#what-features-are-important-for-disease-outcome-prediction"></a>
</h3>
<p>In classifyR, the default settings perform 20 repeats of 5-fold
cross-validation, giving rise to a total of 100 models. Therefore, to
answer what features are important in the modelling, we first need to
examine whether the rankings (or the importance) of each feature in each
model are the same.</p>
<p>We use the feature type “gene_mean_celltype” as an example to
illustrate. Here, we pick one of the cell type specific genes
<em>CREB3</em> and look at its ranks across the 100 models. (The smaller
the rank, the better. Rank of 1 indicates that the feature is the most
important for the model.)</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">gene_mean_celltype</span> <span class="op">&lt;-</span> <span class="va">classifyr_result</span><span class="op">[[</span><span class="st">"gene_mean_celltype.DLDA.t-test"</span>  <span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">Plasma_CREB3_rank</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">thismodel</span> <span class="op">&lt;-</span> <span class="va">gene_mean_celltype</span><span class="op">@</span><span class="va">rankedFeatures</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="va">thismodel</span><span class="op">$</span><span class="va">rank</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">thismodel</span><span class="op">)</span></span>
<span>    <span class="va">thismodel</span><span class="op">[</span><span class="st">"Plasma..CREB3"</span>, <span class="op">]</span><span class="op">$</span><span class="va">rank</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Plasma_CREB3_rank</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> rank <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">Plasma_CREB3_rank</span><span class="op">)</span> <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">Plasma_CREB3_rank</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span> x <span class="op">=</span> <span class="st">"model"</span>, y <span class="op">=</span> <span class="va">rank</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> </span></code></pre></div>
<p><img src="disease_outcome_classification_schulte_files/figure-html/unnamed-chunk-13-1.png" width="288"></p>
<p>This feature type contains a total of 26076 features. Therefore,
although the boxplot shows a certain variability in the ranks of the
features, overall it is consistent that the expression of <em>CREB3</em>
in plasma cells has an important contribution towards the classification
model.</p>
<p>This then leads to the next question, how should we determine the
final set of top features to focus on?</p>
<p>Here, we use the mean scores across the folds to calculate the final
set of top features.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_features</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">thismodel</span> <span class="op">&lt;-</span> <span class="va">gene_mean_celltype</span><span class="op">@</span><span class="va">rankedFeatures</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="va">thismodel</span><span class="op">$</span><span class="va">rank</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">thismodel</span><span class="op">)</span></span>
<span>    <span class="va">thismodel</span> <span class="op">&lt;-</span> <span class="va">thismodel</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">thismodel</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span>    <span class="va">thismodel</span><span class="op">$</span><span class="va">model</span> <span class="op">&lt;-</span> <span class="va">x</span></span>
<span>    <span class="va">thismodel</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">top_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">top_features</span> <span class="op">)</span> <span class="op">)</span> </span>
<span></span>
<span><span class="va">top_features</span> <span class="op">&lt;-</span> <span class="va">top_features</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">feature</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>rank <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">rank</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">rank</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">top_features</span>,<span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 2</span></span></span>
<span><span class="co">#&gt;    feature              rank</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Neutrophil..CD63     5.88</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Neutrophil..S100A12 12.4 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Neutrophil..TSPO    12.7 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Neutrophil..S100P   13.3 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Neutrophil..MYL6    16.6 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> CD4.T..LDHA         17.3 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> CD4.T..GAPDH        20.8 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Neutrophil..TOMM7   21.4 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Neutrophil..CYBB    21.4 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Neutrophil..PRDX5   25.0</span></span></code></pre></div>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<p>Further explore the ranks of the features. Which cell type(s) would
you like to focus on at your next stage of analysis ?</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="is-the-constructed-model-robust-for-disease-outcome-classification">5. Is the constructed model robust for disease outcome
classification?<a class="anchor" aria-label="anchor" href="#is-the-constructed-model-robust-for-disease-outcome-classification"></a>
</h2>
<p>In this section we take a detailed look at the model performance. In
particular, a number of factors may affect the classification
performance, such as the choice of classification algorithm.</p>
<div class="section level3">
<h3 id="do-the-results-change-with-different-classification-algorithms">5.1 Do the results change with different classification
algorithms?<a class="anchor" aria-label="anchor" href="#do-the-results-change-with-different-classification-algorithms"></a>
</h3>
<p>classifyR provides an implementation of a number of commonly used
classification algorithms.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">classifyr_result_svm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ClassifyR/man/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span><span class="va">scfeatures_result</span>,</span>
<span>                            <span class="va">outcome</span>, </span>
<span>                            classifier <span class="op">=</span> <span class="st">"SVM"</span>,</span>
<span>                            nCores <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p>Compare the accuracy obtained from the different algorithms.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="va">format_accuracy</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">result_list</span>, <span class="va">performanceType</span> <span class="op">=</span> <span class="st">"Balanced Accuracy"</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">result_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">result_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/ClassifyR/man/calcPerformance.html" class="external-link">calcCVperformance</a></span><span class="op">(</span><span class="va">x</span>, performanceType <span class="op">=</span> <span class="va">performanceType</span> <span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">accuracy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">result_list</span>, <span class="va">performance</span><span class="op">)</span></span>
<span>  <span class="va">accuracy</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">accuracy</span> , <span class="va">`[`</span>, <span class="va">performanceType</span><span class="op">)</span></span>
<span>  <span class="va">accuracy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span> <span class="va">accuracy</span>  <span class="op">)</span></span>
<span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">accuracy</span> <span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span> <span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">result_list</span><span class="op">)</span> , <span class="st">"\\."</span> <span class="op">)</span>, <span class="va">`[`</span> , <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">accuracy</span> <span class="op">&lt;-</span> <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">accuracy</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">accuracy</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://www.dropbox.com/s/43c72sqcxnwnjpi/classifyr_result_SVM.rds?raw=1"</span> , destfile <span class="op">=</span> <span class="st">"intermediate_result/classifyr_result_SVM.rds"</span> , method<span class="op">=</span><span class="st">"auto"</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">classifyr_result_SVM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"intermediate_result/classifyr_result_SVM.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">accuracy_SVM</span> <span class="op">&lt;-</span> <span class="fu">format_accuracy</span><span class="op">(</span><span class="va">classifyr_result_SVM</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">accuracy_SVM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">accuracy_SVM</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">variable</span>, y <span class="op">=</span> <span class="va">value</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Feature types"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Balanced Accuracy - SVM"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">accuracy_SVM</span></span></code></pre></div>
<p><img src="disease_outcome_classification_schulte_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span> <span class="co"># ggplot(accuracy_SVM, aes(x = variable, y = value, colour = variable )) + geom_boxplot() + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) + xlab("Feature types") + ylab("Balanced Accuracy - SVM") + scale_colour_tableau("Tableau 20")+ ylim(0,1)</span></span>
<span></span>
<span></span>
<span><span class="co"># ggsave("accuracy_on_schulte.pdf", width = 16, height = 10, units= "cm")</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="do-the-results-change-with-different-evaluation-metrics">5.2 Do the results change with different evaluation metrics?<a class="anchor" aria-label="anchor" href="#do-the-results-change-with-different-evaluation-metrics"></a>
</h3>
<p>Here we calculate accuracy instead of balanced accuracy. Inspect the
difference in performance.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="va">classifyr_result_accuracy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">classifyr_result_SVM</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/ClassifyR/man/calcPerformance.html" class="external-link">calcCVperformance</a></span><span class="op">(</span><span class="va">x</span>, performanceType <span class="op">=</span> <span class="st">"Accuracy"</span> <span class="op">)</span><span class="op">)</span></span>
<span> </span>
<span> </span>
<span></span>
<span><span class="va">accuracy_SVM_accuracy</span> <span class="op">&lt;-</span> <span class="fu">format_accuracy</span><span class="op">(</span><span class="va">classifyr_result_accuracy</span><span class="op">)</span></span>
<span></span>
<span><span class="va">accuracy_SVM_accuracy</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">accuracy_SVM_accuracy</span> , <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">variable</span>, y <span class="op">=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Feature types"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Balanced Accuracy - SVM Accuracy"</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">accuracy_SVM_accuracy</span> </span></code></pre></div>
<p><img src="disease_outcome_classification_schulte_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="is-the-model-generalisable">5.3 Is the model generalisable?<a class="anchor" aria-label="anchor" href="#is-the-model-generalisable"></a>
</h3>
<p>Generalisable means whether the model can have good performance when
tested on an independent dataset.</p>
<p>Here, we use the model built on the Schulte-Schrepping dataset to
test on the Wilk dataset. The Wilk data is obtained from <a href="https://rupress.org/jem/article/218/8/e20210582/212379/Multi-omic-profiling-reveals-widespread" class="external-link">Wilk
et al. 2022</a>. It contains 23 mild and 19 severe patients.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://www.dropbox.com/s/6ecpzf2qrnglayg/wilk_full.rds?raw=1"</span> , destfile <span class="op">=</span> <span class="st">"toy_data/wilk_full.rds"</span> , method<span class="op">=</span><span class="st">"auto"</span> <span class="op">)</span></span>
<span> </span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"toy_data/wilk_full.rds"</span><span class="op">)</span></span></code></pre></div>
<p>First visualise the data.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span> <span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">data</span> <span class="op">)</span>  </span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">data</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">data</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">data</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 112589</span></span>
<span><span class="co">#&gt; Number of edges: 3319297</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.9204</span></span>
<span><span class="co">#&gt; Number of communities: 19</span></span>
<span><span class="co">#&gt; Elapsed time: 88 seconds</span></span>
<span><span class="va">data</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">data</span> , dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">data</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"meta_severity"</span><span class="op">)</span></span></code></pre></div>
<p><img src="disease_outcome_classification_schulte_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>Here we have provided the constructed scFeatures result for the Wilk
dataset.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">scfeatures_result_schulte</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"intermediate_result/scfeatures_result_schulte.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://www.dropbox.com/s/wihtyecqmouf22a/scfeatures_result_wilk_full.rds?raw=1"</span> , destfile <span class="op">=</span> <span class="st">"intermediate_result/scfeatures_result_wilk_full.rds"</span> , method<span class="op">=</span><span class="st">"auto"</span> <span class="op">)</span></span>
<span></span>
<span> </span>
<span></span>
<span><span class="va">scfeatures_result_wilk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"intermediate_result/scfeatures_result_wilk_full.rds"</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">scfeatures_result_wilk</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result_wilk</span><span class="op">[</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result_schulte</span> <span class="op">)</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="co"># in order to train and test two different datasets, need to make sure the features are the same </span></span>
<span><span class="co"># pick the common features </span></span>
<span><span class="va">new_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">scfeatures_result_schulte</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>   <span class="va">common_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span> <span class="va">scfeatures_result_schulte</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> ,</span>
<span>             <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span> <span class="va">scfeatures_result_wilk</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>   <span class="va">new_schulte</span> <span class="op">&lt;-</span>  <span class="va">scfeatures_result_schulte</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="va">common_features</span><span class="op">]</span></span>
<span>   <span class="va">new_wilk</span> <span class="op">&lt;-</span> <span class="va">scfeatures_result_wilk</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="va">common_features</span><span class="op">]</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">new_schulte</span>, <span class="va">new_wilk</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># format the features </span></span>
<span><span class="va">new_scfeatures_result_schulte</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">new_list</span>, <span class="va">`[[`</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">new_scfeatures_result_wilk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">new_list</span>, <span class="va">`[[`</span>, <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">new_scfeatures_result_schulte</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result_schulte</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">new_scfeatures_result_wilk</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">scfeatures_result_wilk</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># First clean the column names</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">new_scfeatures_result_schulte</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">new_scfeatures_result_schulte</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\s|[[:punct:]]"</span>, <span class="st">"."</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">new_scfeatures_result_schulte</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># obtain the outcome, which is stored in the rownames of ech matrix </span></span>
<span><span class="va">outcome_schulte</span> <span class="op">=</span> <span class="va">scfeatures_result_schulte</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">rownames</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="st">"_cond_"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">outcome_wilk</span> <span class="op">=</span> <span class="va">scfeatures_result_wilk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">rownames</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="st">"_cond_"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">outcome_wilk</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">scfeatures_result_wilk</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># lapply(new_scfeatures_result_schulte, dim)</span></span>
<span><span class="co"># lapply(new_scfeatures_result_wilk, dim)</span></span>
<span></span>
<span><span class="co"># remove the feature types that have no common features </span></span>
<span><span class="co"># new_scfeatures_result_schulte[[6]] &lt;- NULL</span></span>
<span><span class="co"># new_scfeatures_result_wilk[[6]] &lt;- NULL</span></span></code></pre></div>
<p>First, we use the dataset from Schulte-Schrepping to perform self
cross-validation. The purpose is to identify the best model (the best
parameters). Once we decide on the best model, we then use this model to
test on the Wilk dataset.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># perform self cross validation</span></span>
<span><span class="va">classifyr_result_svm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ClassifyR/man/crossValidate.html" class="external-link">crossValidate</a></span><span class="op">(</span><span class="va">new_scfeatures_result_schulte</span>,</span>
<span>                            <span class="va">outcome_schulte</span>, </span>
<span>                            classifier <span class="op">=</span> <span class="st">"SVM"</span>,</span>
<span>                            nCores <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate balanced accuracy </span></span>
<span></span>
<span><span class="va">classifyr_result_svm</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">classifyr_result_svm</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/ClassifyR/man/calcPerformance.html" class="external-link">calcCVperformance</a></span><span class="op">(</span><span class="va">x</span>, performanceType <span class="op">=</span> <span class="st">"Balanced Accuracy"</span> <span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># for each feature type, identify the best model and use the best model to predict on Wilk </span></span>
<span></span>
<span><span class="va">result_generalisability</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="kw">for</span><span class="op">(</span> <span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span> <span class="va">classifyr_result_svm</span> <span class="op">)</span><span class="op">)</span> <span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">bestmodel</span> <span class="op">&lt;-</span> <span class="va">classifyr_result_svm</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">finalModel</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">tuneList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>cost <span class="op">=</span>  <span class="va">bestmodel</span><span class="op">$</span><span class="va">cost</span>, </span>
<span>                   gamma <span class="op">=</span> <span class="va">bestmodel</span><span class="op">$</span><span class="va">gamma</span>, </span>
<span>                   degree <span class="op">=</span> <span class="va">bestmodel</span><span class="op">$</span><span class="va">degree</span>, </span>
<span>                   type <span class="op">=</span> <span class="va">bestmodel</span><span class="op">$</span><span class="va">type</span>,</span>
<span>                   tot.nSV<span class="op">=</span> <span class="va">bestmodel</span><span class="op">$</span><span class="va">tot.nSV</span><span class="op">)</span></span>
<span>  </span>
<span>   <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/train.html" class="external-link">train</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">new_scfeatures_result_schulte</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> ,outcomeTrain <span class="op">=</span> <span class="va">outcome_schulte</span>,</span>
<span>                  performanceType <span class="op">=</span> <span class="st">"Balanced Accuracy"</span>, classifier <span class="op">=</span> <span class="st">"SVM"</span>,</span>
<span>                  tuneParams <span class="op">=</span> <span class="va">tuneList</span><span class="op">)</span></span>
<span>   </span>
<span>  <span class="va">prediction</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span> <span class="va">model</span> , <span class="va">new_scfeatures_result_wilk</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">truth</span> <span class="op">&lt;-</span>  <span class="va">outcome_wilk</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">prediction</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">temp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ClassifyR/man/calcPerformance.html" class="external-link">calcExternalPerformance</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">truth</span><span class="op">)</span> , <span class="va">prediction</span>, </span>
<span>                                      performanceType <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Balanced Accuracy"</span> <span class="op">)</span><span class="op">)</span> </span>
<span>  <span class="va">temp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>balanced_accuracy <span class="op">=</span> <span class="va">temp</span>, </span>
<span>                     featuretype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">new_scfeatures_result_schulte</span><span class="op">)</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">)</span></span>
<span>    </span>
<span>   <span class="va">result_generalisability</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">result_generalisability</span>,   <span class="va">temp</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># saveRDS( result_generalisability, "intermediate_result/result_list_generalisability.rds")</span></span>
<span> </span></code></pre></div>
<p>Visualise the accuracy using boxplots.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://www.dropbox.com/s/9fmyvw9z1eh47ps/result_list_generalisability.rds?raw=1"</span> , destfile <span class="op">=</span> <span class="st">"intermediate_result/result_list_generalisability.rds"</span> , method<span class="op">=</span><span class="st">"auto"</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span> <span class="va">result_list_generalisability</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"intermediate_result/result_list_generalisability.rds"</span><span class="op">)</span></span>
<span></span>
<span> <span class="va">result_list_generalisability</span><span class="op">$</span><span class="va">balanced_accuracy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span> <span class="va">result_list_generalisability</span><span class="op">$</span><span class="va">balanced_accuracy</span>, <span class="fl">2</span><span class="op">)</span></span>
<span> </span>
<span> </span>
<span> <span class="va">result_list_generalisability</span><span class="op">$</span><span class="va">featuretype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span>  <span class="va">result_list_generalisability</span><span class="op">$</span><span class="va">featuretype</span> , levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="va">result_list_generalisability</span><span class="op">$</span><span class="va">featuretype</span><span class="op">)</span><span class="op">)</span></span>
<span> </span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span> <span class="va">result_list_generalisability</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">featuretype</span> , y <span class="op">=</span> <span class="va">balanced_accuracy</span>, fill <span class="op">=</span> <span class="va">featuretype</span> <span class="op">)</span> <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_col</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/scale_color_tableau.html" class="external-link">scale_fill_tableau</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Tableau 20"</span><span class="op">)</span> <span class="op">+</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">balanced_accuracy</span><span class="op">)</span>, vjust <span class="op">=</span> <span class="op">-</span><span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="disease_outcome_classification_schulte_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> </span>
<span><span class="co"># ggsave("accuracy_on_wilk.pdf" , width = 16, height = 10, units = "cm")</span></span></code></pre></div>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Discussion:</strong></p>
<p>Examine the classification accuracy and comment on the
generalisability of the model.</p>
</div>
<div class="aimbox">
<p><span class="fa-stack fa-lg">
<i class="fa fa-circle fa-stack-2x"></i> <span class="far fa-comment fa-stack-1x fa-inverse"></span> </span>
<strong>Final discussion:</strong></p>
<p>How good do you think our model is? What parts of the workflow could
you change to improve the results?</p>
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Sydney Precision Bioinformatics Alliance.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
